Source/HTTP/iOS/http_ios.mm

```diff
@@ -0,0 +1,134 @@
+
+    uint32_t timeoutInSeconds = 0;
+    if (FAILED(HCHttpCallRequestGetTimeout(m_call, &timeoutInSeconds)))
+    {
+        // default to 60 to match other default ios behaviour
+        timeoutInSeconds = 60;
+    }
+    
+    [configuration setTimeoutIntervalForRequest:(NSTimeInterval)timeoutInSeconds];
+    [configuration setTimeoutIntervalForResource:(NSTimeInterval)timeoutInSeconds];
+    
+    m_session = [NSURLSession sessionWithConfiguration:configuration];
+}
+
+void ios_http::completion_handler(NSData *data, NSURLResponse *response, NSError *error)
+{
+    ASSERT([response isKindOfClass:[NSHTTPURLResponse class]]);
+    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)response;
+    
+    if (error)
+    {
+        // TODO: handle errors
+        CompleteAsync(m_asyncBlock, E_FAIL, 0);
+        return;
+    }
+
+    uint32_t statusCode = (uint32_t)[httpResponse statusCode];
+    
+    HCHttpCallResponseSetStatusCode(m_call, statusCode);
+    
+    NSDictionary *headers = [httpResponse allHeaderFields];
+    for (NSString *key in headers)
+    {
+        NSString *value = headers[key];
+        
+        const char *keyCString = [key cStringUsingEncoding:NSUTF8StringEncoding];
+        const char *valueCString = [value cStringUsingEncoding:NSUTF8StringEncoding];
+        HCHttpCallResponseSetHeader(m_call, keyCString, valueCString);
+    }
+
+    HCHttpCallResponseSetResponseBodyBytes(m_call, static_cast<const uint8_t*>([data bytes]), [data length]);
+    CompleteAsync(m_asyncBlock, S_OK, 0);
+}
+
+void ios_http::initiate_request()
+{
+    const char* urlCString = nullptr;
+    const char* methodCString = nullptr;
+    if (FAILED(HCHttpCallRequestGetUrl(m_call, &methodCString, &urlCString)))
+    {
+        CompleteAsync(m_asyncBlock, E_FAIL, 0);
+        return; //TODO: handle errors better
+    }
+    
+    NSString* urlString = [[NSString alloc] initWithUTF8String:urlCString];
+    NSURL* url = [NSURL URLWithString:urlString];
+    
+    NSString* methodString = [[NSString alloc] initWithUTF8String:methodCString];
+    
+    NSMutableURLRequest* request = [NSMutableURLRequest requestWithURL:url];
+    [request setHTTPMethod:methodString];
+    
+    uint32_t numHeaders = 0;
+    if (FAILED(HCHttpCallRequestGetNumHeaders(m_call, &numHeaders)))
+    {
+        CompleteAsync(m_asyncBlock, E_FAIL, 0);
+        return; //TODO: errors
+    }
+    
+    for(uint32_t i=0; i<numHeaders; ++i)
+    {
+        const char* headerName;
+        const char* headerValue;
+        if (SUCCEEDED(HCHttpCallRequestGetHeaderAtIndex(m_call, i, &headerName, &headerValue)))
+        {
+            NSString* headerNameString = [[NSString alloc] initWithUTF8String:headerName];
+            NSString* headerValueString = [[NSString alloc] initWithUTF8String:headerValue];
+            
+            [request addValue:headerValueString forHTTPHeaderField:headerNameString];
+        }
+    }
+    
+    const BYTE* requestBody = nullptr;
```

Comment:
BYTE const*

Reply:
uint8_t const* actually

---
In reply to: [188698957](https://github.com/Microsoft/libHttpClient/pull/162#discussion_r188698957) [](ancestors = 188698957)